robot_smach_states/benchmark/benchmark_grasping.py

```diff
@@ -0,0 +1,184 @@
+                        help="Only grasp an item if it matches the specified class. "
+                             "Pass --non-strict-class=False to also grab another class of items "
+                             "if none of the expected class is observed. "
+                             "Use this if you only care about grasping, not object recognition")
+
+    subparsers = parser.add_subparsers(help='single or batch mode', dest='subcommand')
+    single = subparsers.add_parser(name='single', description="Measure the time it takes to grasp an entity when starting from a given pose. "
+                                                 "After the grasping is done, go back to the waypoint, "
+                                                 "turn around and drop the item, "
+                                                 "then turn around once more to be in the start configuration again")
+    single.add_argument("cls", type=str,
+                        help="class of entity to grasp, eg. 'coke' or '*' if you don't care")
+    single.add_argument("support", type=str,
+                        help="ID of entity to grasp FROM ('grasp' entity is on-top-of this 'support' entity), eg. 'cabinet'")
+    single.add_argument("waypoint", type=str,
+                        help="ID of waypoint for start and end position of the robot")
+    single.add_argument("--inspect_from_area", type=str,
+                        help="What area of the support-entity to inspect from", default='in_front_of')
+
+
+    batch = subparsers.add_parser(name='batch', description="Perform the single case repeatedly, "
+                                                           "taking configurations from a .csv file, "
+                                                            "with columns 'cls,support,waypoint,inspect_from_area' (the latter may be empty)")
+    batch.add_argument("configuration", type=str, default='grasp_benchmark_config.csv')
+
+    args = single.parse_args()
+    rospy.init_node("benchmark_grasping")
+
+    with open(args.output, 'a+') as csv_file:
+        reader = csv.DictReader(csv_file)
+
+        fields = ['robot',
+                  'start_waypoint',
+                  'expected_class', 'observed_class',
+                  'id',
+                  'inspect_duration',
+                  'grab_duration',
+                  'x', 'y', 'z']
+        results_writer = csv.DictWriter(csv_file, fieldnames=fields, extrasaction='ignore')  # Ignore extra keys in the dicts
+
+        if reader.fieldnames != results_writer.fieldnames:
+            results_writer.writeheader()
+
+        robot = get_robot(args.robot)
+
+        if args.subcommand == 'single':
+            single_item(robot, results_writer,
+                        args.cls, args.support, args.waypoint, args.inspect_from_area,
+                        args.non_strict_class)
+        elif args.subcommand == 'batch':
+            with open(args.configuration) as config_file:
```

Comment:
TBH I can only partly follow this read/write shizzle, but I think a config_file is needed, which is not available on git, so that needs to be added

Reply:
Yes, hope to make that tomorrow